internal/federationdomain/resolvedprovider/resolvedgithub/resolved_github_provider.go

```diff
@@ -91,18 +95,80 @@ func (p *FederationDomainResolvedGitHubIdentityProvider) Login(
 }
 
 func (p *FederationDomainResolvedGitHubIdentityProvider) LoginFromCallback(
-	_ context.Context,
-	_ string,
-	_ pkce.Code,
-	_ nonce.Nonce,
-	_ string,
+	ctx context.Context,
+	authCode string,
+	_ pkce.Code, // GitHub does not support PKCE, see https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps
+	_ nonce.Nonce, // GitHub does not support OIDC, therefore there is no ID token that could contain the "nonce".
+	redirectURI string,
 ) (*resolvedprovider.Identity, *resolvedprovider.IdentityLoginExtras, error) {
-	return nil, nil, errors.New("function LoginFromCallback not yet implemented for GitHub IDP")
+	accessToken, err := p.Provider.ExchangeAuthcode(ctx, authCode, redirectURI)
+	if err != nil {
+		plog.WarningErr("error exchanging GitHub authcode", err, "upstreamName", p.Provider.GetName())
+		return nil, nil, httperr.Wrap(http.StatusBadGateway,
+			fmt.Sprintf("failed to exchange authcode using GitHub API: %s", err.Error()),
+			err,
+		)
+	}
+
+	user, err := p.Provider.GetUser(ctx, accessToken, p.GetDisplayName())
+	if err != nil {
+		return nil, nil, httperr.Wrap(http.StatusUnprocessableEntity,
+			fmt.Sprintf("failed to get user info from GitHub API: %s", err.Error()),
+			err,
+		)
+	}
+
+	return &resolvedprovider.Identity{
+			UpstreamUsername:  user.Username,
+			UpstreamGroups:    user.Groups,
+			DownstreamSubject: user.DownstreamSubject,
+			IDPSpecificSessionData: &psession.GitHubSessionData{
+				UpstreamAccessToken: accessToken,
+			},
+		},
+		&resolvedprovider.IdentityLoginExtras{
+			DownstreamAdditionalClaims: nil, // not using this for GitHub
+			Warnings:                   nil, // not using this for GitHub
+		},
+		nil // no error
 }
 
 func (p *FederationDomainResolvedGitHubIdentityProvider) UpstreamRefresh(
```

Comment:
I implemented upstream refresh for GitHubIdentityProvider and added an integration test for it.

TODO: Add a few unit tests to `token_handler_test.go` to show that the token endpoint works with GitHubIdentityProvider. Shouldn't need to change its production code. Similar to the callback handler units tests, can cover basic things like:
- Can do an authcode exchange
- Can do a refresh
- The downstream refresh fails when the mock upstream refresh fails
- Maybe also similar cases for when using a dynamic client?

Reply:
Done in https://github.com/vmware-tanzu/pinniped/pull/1963/commits/bb9cb739c655cfe8ef0dcef40decf27e2299a77d